geo/src/algorithm/stitch.rs

```diff
@@ -0,0 +1,581 @@
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+                $($impls)*
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<&$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.as_slice().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for &[$type]
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
+           }
+       }
+
+       impl<T> Stitch<T> for Vec<$type>
+       where
+           T: GeoFloat,
+       {
+           fn stitch_together(&$self) -> PolygonStitchingResult<MultiPolygon<T>> {
+               $self.iter().collect::<Vec<_>>().stitch_together()
+           }
+       }
+     };
+}
+
+impl_stitch! {
+    Polygon<T> => self
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+        let polys = self
+            .iter()
+            .map(|&poly| fix_orientation(poly.clone()))
+            .collect::<Vec<_>>();
+        let lines = polys
+            .iter()
+            .flat_map(|part| part.lines_iter())
+            .collect::<Vec<_>>();
+
+        let boundary_lines = find_boundary_lines(lines);
+        let stitched_multipolygon = stitch_multipolygon_from_lines(boundary_lines)?;
+
+        let polys = stitched_multipolygon
+            .into_iter()
+            .map(find_and_fix_holes_in_exterior)
+            .collect::<Vec<_>>();
+
+        Ok(MultiPolygon::new(polys))
+    }
+}
+
+impl_stitch! {
+    Triangle<T> => self
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+        self.iter()
+            .map(|tri| tri.to_polygon())
+            .collect::<Vec<_>>()
+            .stitch_together()
+    }
+}
+
+impl_stitch! {
+    MultiPolygon<T> => self
+    fn stitch_together(&self) -> PolygonStitchingResult<MultiPolygon<T>> {
+        self.iter().flat_map(|mp| mp.0.iter()).collect::<Vec<_>>().stitch_together()
+    }
+}
+
+/// makes interiors and exteriors of polygon have ccw orientation
+fn fix_orientation<T: GeoFloat>(mut poly: Polygon<T>) -> Polygon<T> {
+    poly.exterior_mut(|ls| ls.make_ccw_winding());
+
+    // just commenting this out for the moment. The algorithm only works on exteriors so this
+    // shouldn't be needed
+    //
+    //poly.interiors_mut(|ls| ls.iter_mut().for_each(|ls| ls.make_ccw_winding()));
+
+    poly
+}
+
+/// checks whether the to lines are equal or inverted forms of each other
+#[inline]
+fn same_line<T: GeoFloat>(l1: &Line<T>, l2: &Line<T>) -> bool {
+    (l1.start == l2.start && l1.end == l2.end) || (l1.start == l2.end && l2.start == l1.end)
+}
+
+/// given a collection of lines from multiple polygons, this returns all but the shared lines
+fn find_boundary_lines<T: GeoFloat>(lines: Vec<Line<T>>) -> Vec<Line<T>> {
```

Comment:
I confess, I don't really understand the purpose of this method, but I think you changed the behavior with your recent optimization in a7b9b7f7334d95bf8daa71a2e27087e70554f5b1

For example, consider these two methods which work on ints:
```
uniq(1, 1, 2, 4) == { 1, 2, 4 }
```

```
no_repeats(1, 1, 2, 4) ==  {2, 4}
```

Previously this method worked like `no_repeats`, but now it works like `uniq`. Right? 


Are you trying to return *unique* line segments? Or are you trying to return all methods which have no duplicates?

The distinction

Reply:
It'll work like this now:

```rust
- input `([], 1)`
  - `1` is not in `[]`
  - push `1` to `[]`
- input `([1], 1)`
  - `1` is in `[1]` as idx `0`
  - remove elem at idx `0` from `[1]`
  - don't push input `1`
- ...
- pushes `2` and `4`
```

To be fair, even though it works in this scenario, you're still right. The algo changes. It will deduplicate pair-wise equal elements. What does that mean? The following inputs produce the respective outputs

- `f(1, 2, 4) == {1, 2, 4}`
- `f(1, 1, 2, 4) == {2, 4}`
- `f(1, 1, 1, 2, 4) == {1, 2, 4}`
- ...
- even amounts of `1`s -> no `1`s in result
- uneven amounts of `1`s -> exactly one `1` in result

I would argue that this still is ok for this algo (I should definitely leave a comment though). The explanation is that the algorithm is supposed to only run on a  intersection less partition of a bounded 2d area, e.g. a triangulation of a polygon. In this case we only have the following two possibilities:

1. The line is a boundary line -> it's unique
2. The line is a inner line of the partition -> it has exactly one duplicate from an adjacent polygon

I could add some extra logic to prevent duplicate lines even if we get an uneven amount. That also shouldn't add too much to the performance overhead, though.

---

Writing all of this, I come to the conclusions:

- Maybe we should place more checks before running the actual algo and return descriptive errors that tell the users if a precondition is not met
- It's enough for SpadeBoolops to have the stitching of a triangulation. Maybe we shouldn't generalize the algorithm so much to work on polygons in the first place since the input assumptions are very special and I fear that users will get confused

Reply:
Ah! When I think about this algorithm in terms of stitching together a triangulation, it becomes much more clear to me why this edge-removal stuff is supposed to work. 

I agree that it's not  a good idea to not over generalize, and I think we'd have an easier time supporting an algorithm that works on triangles only for now. (As a bonus, triangles can be stack allocated, so we might be able to avoid a lot of the vec shuffling.)

A triangle-only stitching algorithm would still support everything you need in #1089 right?

Reply:
Special casing the algo gave us a good performance boost again since we can make a few more assumptions.

Although it's a bit sad that we can't use the algorithm on polygons/multi polygons directly anymore, I think it's an ok workaround to just triangulate and then stitch collections of these. This will hopefully trade off equally in the end with the performance boost gained here from the specialization.